processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java

```diff
@@ -90,14 +120,40 @@
   // Maximum number of elements in the table before it must be resized
   private int maxSize;
 
+  // Limit to apply to results.
+  // If limit > 0, track hash table entries in a binary heap with size of limit.
+  // If -1, no limit is applied, hash table entry offsets are tracked with an unordered list with no limit.
+  private int limit;
```

Comment:
IMO, it would be better to make a new Grouper supporting limit push down because its behavior is quite different from the original one in many parts to achieve a different goal. It will be helpful for better readability and maybe slight performance as well.